### PR TITLE
Adding Stella as A2600 emulation core

### LINKED ISSUE #2001: Add Stella to Bizhawk for Atari 2600 emulation.
I'll be honest, I don't know much about this.  I've heard our 2600 isn't so great and theirs is.

@adelikat I think you've looked into this, could you give a short summary?

### PR DESCRIPTION
**Pull request by SergioMartin86**
[//]: # "This description supports Markdown syntax. There's a cheatsheet here: https://guides.github.com/features/mastering-markdown/"
[//]: # "These lines are comments, for letting you know what you should be writing. You can delete them or leave them in."
[//]: # "Also, please remember to link related Issues! If a bug hasn't been reported, you may submit a fix without creating an Issue."

Adding Stella as A2600 emulator. 

Progress has been painful and messy; don't expect minimalism during this first stage. We can do the fine tuning once everything works correctly.

[//]: # "Apart from the mandatory license signature, these tasks are optional, but doing them could save reviewers some time and get the PR merged sooner."
Check if completed:
- [x] I have run any relevant test suites
- [x] I, the committer, have read the [licensing terms for contributors](https://github.com/TASEmulators/BizHawk/blob/master/contributing.md#copyrights-and-licensing) (last updated 2024-03-20) and am compliant

- [x] Load Rom
- [x] Advancing Frame
- [x] Processing Digital Inputs
- [ ]  Processing Analog Inputs
- [x] Rendering frame into the screen
- [x] Rendering sound
- [x] Memory Accesses
- [ ] CPU Hooks

resolves #2001

### CODE REVIEW

**Comment by CasualPokePlayer** [comment]
Note, GPGX isn't particularly the best example of how to do a waterbox core, if it's at all possible use the standard `WaterboxCore` base class (see PicoDrive, TIC80, uzem, as simple examples of this being used)

**Comment by SergioMartin86** [comment]
Ok, opening this PR for review. This is a summary of what's been done:

- Emulation works correctly for several games I tested (including Myst, which apparently a2600hawk does not support)
- Video is rendering correctly
- Audio is also playing correctly, except for the first few frames or so (try Frogger (USA).a26)
- Statable interface correctly implemented
- Memory domains correctly showing
- Digital Controllers working

To-Dos:

- Analog controllers (paddle) not currently supported
- CPU hooks and debug functonality not current implemented

File: waterbox/stella/BizhawkInterface.cxx [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+#include "MediaFactory.hxx"
+#include "Serializer.hxx"
+#include "StateManager.hxx"
+#include "Console.hxx"
+#include "Control.hxx"
+#include "Switches.hxx"
+#include "M6532.hxx"
+#include "TIA.hxx"
+
+uint16_t soundbuffer[4096];
```
Comment (nattthebear) [inline waterbox/stella/BizhawkInterface.cxx:16]: Can this be placed in the invisible section?

Comment (SergioMartin86) [reply waterbox/stella/BizhawkInterface.cxx:16]: Yes, that'd make sense

**Comment by vadosnaprimer** [comment]
I think this can be merged (after conflicts are fixed) and completed later in master.

**Comment by SergioMartin86** [comment]
* Conflicts have been resolved
* Waterboxed core added
* Fully working for discrete inputs

image

File: waterbox/stella/BizhawkInterface.cxx (hunk 1 above)
Comment (vadosnaprimer) [reply waterbox/stella/BizhawkInterface.cxx:16]: Was this resolved?

Comment (YoshiRulz) [reply waterbox/stella/BizhawkInterface.cxx:16]: This array hasn't moved so I'd say no.

Comment (SergioMartin86) [reply waterbox/stella/BizhawkInterface.cxx]: should be resolved now

File: .vscode/settings.json [hunk 2]
```diff
@@ -1,7 +0,0 @@
-{
-	"editor.insertSpaces": false,
-	"editor.tabSize": 4,
-	"search.exclude": {
-		"waterbox/**": true
-	}
-}
```
Comment (YoshiRulz) [inline .vscode/settings.json:7]: Is this necessary for using VS Code? It shouldn't be included in this PR either way.

Comment (SergioMartin86) [reply .vscode/settings.json:7]: Yeah, these sneaked into the repo without me noticing. Removed.

**Review by CasualPokePlayer — DISMISSED** [review]
_byteArrayDomains is unused (I assume this is copy paste from a2600hawk)

Also, probably should ensure the `Main RAM` domain is the IMemoryDomain's `MainMemory`.

**Comment by SergioMartin86** [comment]
> _byteArrayDomains is unused (I assume this is copy paste from a2600hawk)
> 
> Also, probably should ensure the `Main RAM` domain is the IMemoryDomain's `MainMemory`.

I'm not sure exactly what changes are being requested here

File: .vscode/settings.json (hunk 2 above)
Comment (YoshiRulz) [reply .vscode/settings.json:7]: No, those files are already checked-in, and you've deleted them. If you restore them, does VS Code stop working correctly?

File: src/BizHawk.Emulation.Cores/Consoles/Atari/Stella/Stella.IMemoryDomains.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+using System.Linq;
+
+using BizHawk.Emulation.Common;
+
+namespace BizHawk.Emulation.Cores.Atari.Stella
+{
+	public partial class Stella
+	{
+		internal IMemoryDomains MemoryDomains;
+		private readonly Dictionary<string, MemoryDomainByteArray> _byteArrayDomains = new Dictionary<string, MemoryDomainByteArray>();
```
Comment (YoshiRulz) [inline src/BizHawk.Emulation.Cores/Consoles/Atari/Stella/Stella.IMemoryDomains.cs:11]:
```suggestion
```

File: src/BizHawk.Emulation.Cores/Consoles/Atari/Stella/Stella.IMemoryDomains.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+					_cartMemSize,
+					MemoryDomain.Endian.Little,
+					addr => Core.stella_peek_cartram((uint)addr),
+					(addr, value) => Core.stella_poke_cartram((uint)addr, value),
+					1));
+			}
+
+			domains.Add(new MemoryDomainIntPtrMonitor("Main RAM", MemoryDomain.Endian.Little, mainRamAddress, 128, true, 1, _elf));
+
+			MemoryDomains = new MemoryDomainList(_byteArrayDomains.Values.Concat(domains).ToList());
```
Comment (YoshiRulz) [inline src/BizHawk.Emulation.Cores/Consoles/Atari/Stella/Stella.IMemoryDomains.cs:61]:
```suggestion
			MemoryDomains = new MemoryDomainList(domains) { MainMemory = mainRAM };
```
Requires you save the domain immediately above this to a local `mainRAM`, that is:
```diff
-domains.Add(new MemoryDomainIntPtrMonitor("Main RAM", MemoryDomain.Endian.Little, mainRamAddress, 128, true, 1, _elf));
+MemoryDomainIntPtrMonitor mainRAM = new("Main RAM", MemoryDomain.Endian.Little, mainRamAddress, 128, true, 1, _elf);
+domains.Add(mainRAM);
```

**Review by YoshiRulz — COMMENTED** [review]
These are CPP's requests as committable diffs.

**Comment by SergioMartin86** [comment]
> These are CPP's requests as committable diffs.

Thanks, committed.

File: .vscode/settings.json (hunk 2 above)
Comment (nattthebear) [reply .vscode/settings.json:7]: Those were added in https://github.com/TASEmulators/BizHawk/commit/0b14755ac303533819138a8b7f4aa8f02c780f0c.  This PR shouldn't be modifying them; if we do want to change what we provide for developers and their IDEs, use a separate PR for that.

Comment (SergioMartin86) [reply .vscode/settings.json:7]: Agh I thought it was me who had added them. I never intended to change/add/remove these files. I'll see to put them back to their original state

Comment (SergioMartin86) [reply .vscode/settings.json]: fixed

File: src/BizHawk.Emulation.Cores/Consoles/Atari/Stella/Stella.IMemoryDomains.cs (hunk 4 above)
Comment (SergioMartin86) [reply src/BizHawk.Emulation.Cores/Consoles/Atari/Stella/Stella.IMemoryDomains.cs]: done

File: waterbox/common.mak
Comment (Morilli) [inline waterbox/common.mak:1]: Is there a reason you're doing this instead of just naming the files `*.cpp` instead of `*.cxx`?

**Comment by YoshiRulz** [comment]
Crash before rom finishes loading:
```
[...]
A2600 console created successfully
[Stella] Cart type loaded: 4K (string size: 2, ptr: 3775284406225)
Calling ecl_seal()
WaterboxHost Sealed!
Trace/breakpoint trap (core dumped)
```

File: waterbox/common.mak
Comment (SergioMartin86) [reply waterbox/common.mak:1]: Iirc, there are a lot of .cxx and .hxx files in Stella. Having them all renamed is a much bigger change than simply adding support for them here. Not to mention it would make it very difficult to resync with upstream if all source files are renamed

**Comment by SergioMartin86** [comment]
> Crash before rom finishes loading:
> 
> ```
> [...]
> A2600 console created successfully
> [Stella] Cart type loaded: 4K (string size: 2, ptr: 3775284406225)
> Calling ecl_seal()
> WaterboxHost Sealed!
> Trace/breakpoint trap (core dumped)
> ```

Can you tell me what game you're trying to run?

**Comment by YoshiRulz** [comment]
*Pitfall!* (`SHA1:8D525480445D48CC48460DC666EBAD78C8FB7B73`). Atari2600Hawk loads it with the m4K mapper.

**Review by YoshiRulz — APPROVED** [review]
Approving in this functional state, as per https://github.com/TASEmulators/BizHawk/pull/3911#issuecomment-2269390037.

**Comment by SergioMartin86** [comment]
Pitfall! works now after extending the sound buffer from 4K to 1Mb. Thanks to Morilli for the backtrace needed to find this issue

image

File: waterbox/common.mak
Comment (nattthebear) [reply waterbox/common.mak:1]: I support this change if it's convenient for the core.

File: src/BizHawk.Emulation.Cores/Consoles/Atari/Stella/Stella.cs [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+using BizHawk.Emulation.Common;
+using BizHawk.Emulation.Cores.Consoles.Atari.Stella;
+using BizHawk.Emulation.Cores.Waterbox;
+
+namespace BizHawk.Emulation.Cores.Atari.Stella
+{
+	[PortedCore(
+		name: CoreNames.Stella,
+		author: "The Stella Team",
+//		portedVersion: "", //TODO
```
Comment (YoshiRulz) [inline src/BizHawk.Emulation.Cores/Consoles/Atari/Stella/Stella.cs:15]: This never got filled in
